src/MOI_wrapper.jl

```diff
@@ -0,0 +1,239 @@
+    model.manifold = nothing
+    empty!(model.variable_primal_start)
+    model.sense = MOI.FEASIBILITY_SENSE
+    model.nlp_data = MOI.NLPBlockData([], _EmptyNLPEvaluator(), false)
+    model.qp_data = QPBlockData{Float64}()
+    return nothing
+end
+
+MOI.get(::Optimizer, ::MOI.SolverName) = "Manopt"
+
+MOI.supports_incremental_interface(::Optimizer) = true
+
+function MOI.copy_to(dest::Optimizer, src::MOI.ModelLike)
+    return MOI.default_copy_to(dest, src)
+end
+
+function MOI.supports_add_constrained_variables(::Optimizer, ::Type{<:VectorizedManifold})
+    return true
+end
+
+function MOI.add_constrained_variables(model::Optimizer, set::VectorizedManifold)
+    F = MOI.VectorOfVariables
+    if !isnothing(model.manifold)
+        throw(
+            AddConstraintNotAllowed{F,typeof(set)}(
+                "Only one manifold allowed, variables in `$(model.manifold)` have already been added.",
+            ),
+        )
+    end
+    model.manifold = set.manifold
+    n = MOI.dimension(set)
+    v = MOI.VariableIndex.(1:n)
+    for _ in 1:n
+        push!(model.variable_primal_start, nothing)
+    end
+    return v, MOI.ConstraintIndex{F,typeof(set)}(1)
+end
+
+function MOI.is_valid(model::Optimizer, vi::MOI.VariableIndex)
+    return !isnothing(model.manifold) && 1 <= vi.value <= MOI.dimension(VectorizedManifold(model.manifold))
+end
+
+function MOI.supports(::Optimizer, ::MOI.VariablePrimalStart, ::Type{MOI.VariableIndex})
+    return true
+end
+
+function MOI.set(
+    model::Optimizer,
+    ::MOI.VariablePrimalStart,
+    vi::MOI.VariableIndex,
+    value::Union{Real,Nothing},
+)
+    MOI.throw_if_not_valid(model, vi)
+    model.variable_primal_start[vi.value] = value
+    return nothing
+end
+
+MOI.supports(::Optimizer, ::MOI.NLPBlock) = true
+
+function MOI.set(model::Optimizer, ::MOI.NLPBlock, nlp_data::MOI.NLPBlockData)
+    model.nlp_data = nlp_data
+    return nothing
+end
+
+function MOI.supports(
+    ::Optimizer, ::Union{MOI.ObjectiveSense,MOI.ObjectiveFunction{F}}
+) where {F<:_FUNCTIONS}
+    return true
+end
+
+function MOI.set(model::Optimizer, ::MOI.ObjectiveSense, sense::MOI.OptimizationSense)
+    model.sense = sense
+    return nothing
+end
+
+function MOI.get(
+    model::Optimizer, attr::Union{MOI.ObjectiveFunctionType,MOI.ObjectiveFunction}
+)
+    return MOI.get(model.qp_data, attr)
+end
+
+function MOI.set(
+    model::Optimizer, attr::MOI.ObjectiveFunction{F}, func::F
+) where {F<:_FUNCTIONS}
+    MOI.set(model.qp_data, attr, func)
+    return nothing
+end
+
+function MOI.eval_objective(model::Optimizer, x)
+    if model.sense == MOI.FEASIBILITY_SENSE
+        return 0.0
+    elseif model.nlp_data.has_objective
+        return MOI.eval_objective(model.nlp_data.evaluator, x)
+    end
+    return MOI.eval_objective(model.qp_data, x)
+end
+
+function MOI.eval_objective_gradient(model::Optimizer, grad, x)
+    if model.sense == MOI.FEASIBILITY_SENSE
+        grad .= zero(eltype(y))
```

Comment:
I noticed something more here
(1) what is y? Do you mean x?
(2) the safe way to set grad to be the zero tangent vector at x would be `zero_vector(M, grad, x)` (as soon as you have the problem and called `get_manifold(problem)` to get the manifold.

Reply:
This function is working in the vectorized MOI space while `zero_vector` might give you a matrix etc... ?

Reply:
ah yes, `zero_vector` expects both `grad` and `x` to be in matrix form. The reason for this function is, that while the tangent space (i.e. all _valid_ arrays `grad` can have) is a vector space, but it _might_ happen that the origin is not represented by just the zero vector. Ah and I forgot an `!`. `zero_vector(M, matrix_form(x))` allocates a new matrix for the zero vector, `zero_vector!(M, grad, matrix_form(x))` does that in place of `grad` (assuming it is in matrix form).